### PR TITLE
New login method primarily for jupyter

### PR DESCRIPTION
**Pull request by vanpelt**
This adds a login method and turns on the option for anonymode by default for non-jupyter envs.  Instead of an allow_anonymous arg to init, jupyter users can login anonymously doing `wandb.login(anonymous='allow')`

### CODE REVIEW

File: tests/test_wandb.py [hunk 1]
```diff
@@ -150,6 +150,55 @@ def test_custom_dir(wandb_init_run):
...
+
+def test_login_existing_key(local_netrc):
+    os.environ["WANDB_API_KEY"] = "B" * 40
+    wandb.ensure_configured()
+    wandb.login()
+    assert wandb.api.api_key == "B" * 40
+
+
+def test_login_no_key(local_netrc, mocker):
+    if os.environ.get("WANDB_API_KEY"):
```
Comment (annirudh) [inline tests/test_wandb.py]: Maybe this should be a fixture that we autouse to ensure that the API key is always cleared before these tests?

File: wandb/__init__.py [hunk 2]
```diff
@@ -284,7 +285,34 @@ def load_ipython_extension(ipython):
...
+    """Ensure this machine is logged in
+
+       You can manually specify a key, but this method is intended to prompt for user input.
+
+       anonymous can be "never", "must", or "allow".  If set to "must" we'll always login anonymously,
+       if set to "allow" we'll only create an anonymous user if the user isn't already logged in.
+    """
+    # This ensures we have a global api object
+    ensure_configured()
+    os.environ[env.ALLOW_ANONYMOUS] = anonymous
```
Comment (raubitsj) [inline wandb/__init__.py]: Can you change `env.ALLOW_ANONYMOUS` to `env.ANONYMOUS` (and the underlying env variable to `WANDB_ANONYMOUS`.

File: wandb/__init__.py [hunk 3]
```diff
@@ -284,7 +285,34 @@ def load_ipython_extension(ipython):
...
+            termwarn("Calling wandb.login() without arguments from jupyter should prompt you for an api key.")
+        util.set_api_key(api, key)
+        return key
+    elif api.api_key and anonymous != "must":
+        return api.api_key
+    elif in_jupyter:
+        os.environ[env.JUPYTER] = "true"
+        return _jupyter_login(api=api)
+    else:
+        return util.prompt_api_key(api, force_anonymous=anonymous == "must")
```
Comment (raubitsj) [inline wandb/__init__.py]: Should do a similar check with util.prompt_api_key in `wandb.init()`

**Comment by raubitsj** [comment]
As for the API,  we might want to do something like:
```python
def login(anonymous=None):
  if anonymous:
    os.environment[env.ANONYMOUS] = anonymous
  anonymous = anonymous or "never"
...
def init(anonymous="allow): 
  ...
```

I am just not sure if we want our corporate users to have to turn off anonymous imlicitly by adding a wandb.login() and count on the side effect to disable anonymous.  It seems more transparent to tell them to add something more explicit like:

wandb.login(anonymous="never")
or
wandb.init(anonymous="never")

**Pull request merged**